snuba/utils/streams/batching.py

```diff
@@ -61,6 +62,150 @@ class Offsets:
+    """
+    The ``BatchProcessor`` is a message processor that accumulates processed
+    message values, periodically flushing them after a given duration of time
+    has passed or number of output values have been accumulated.
+    """
+
+    def __init__(
+        self,
+        consumer: Consumer[TPayload],
+        worker: AbstractBatchWorker[TPayload, TResult],
+        max_batch_size: int,
+        max_batch_time: int,
+        metrics: MetricsBackend,
+    ) -> None:
+        self.__consumer = consumer
+        self.__worker = worker
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+        self.__metrics = metrics
+
+        self.__batch: Optional[Batch[TResult]] = None
+        self.__closed = False
+
+    def process(self, message: Optional[Message[TPayload]]) -> None:
+        """
+        Process a message or heartbeat.
+
+        Calling this method may trigger an in-flight batch to be flushed.
+        """
+        assert not self.__closed
+
+        # If we have an active batch, check if it's ready to be flushed.
+        if self.__batch is not None and (
+            len(self.__batch.results) >= self.__max_batch_size
+            or time.time() > self.__batch.created + self.__max_batch_time / 1000.0
+        ):
+            self.__flush()
+
+        # If this was just a heartbeat, there is nothing to process and we can
+        # skip the rest of the method.
+        if message is None:
+            return
+
+        start = time.time()
+
+        self.__metrics.timing(
+            "receive_latency",
+            (start - message.timestamp.timestamp()) * 1000,
+            tags={
+                "topic": message.partition.topic.name,
+                "partition": str(message.partition.index),
+            },
+        )
+
+        # Create the batch only after the first message is seen.
+        if self.__batch is None:
+            self.__batch = Batch()
+
+        result = self.__worker.process_message(message)
+
+        # XXX: ``None`` is indistinguishable from a potentially valid return
+        # value of ``TResult``!
+        if result is not None:
+            self.__batch.results.append(result)
+
+        duration = (time.time() - start) * 1000
+        self.__batch.messages_processed_count += 1
+        self.__batch.processing_time_ms += duration
+        self.__metrics.timing("process_message", duration)
+
+        if message.partition in self.__batch.offsets:
+            self.__batch.offsets[message.partition].hi = message.get_next_offset()
+        else:
+            self.__batch.offsets[message.partition] = Offsets(
+                message.offset, message.get_next_offset()
+            )
+
+    def close(self) -> None:
+        """
+        Close the processor, discarding any messages (without committing
+        offsets) that were previously consumed and processed since the last
+        batch flush.
+        """
+        self.__closed = True
```

Comment:
Note that this behavior changed subtly from the previous implementation. Before, the consumer would discard in-flight messages during rebalancing but flush them during shutdown. This implementation doesn't draw a distinction between the two events, so this discards them in both cases. I don't have a strong opinion about which is "correct" in this case (I think this is too general to make an absolute statement, really) so I'm ok with flushing here instead, but I think it is good to treat both events the same if possible (because it's simpler.)

Reply:
Cannot see a strong reason for doing it differently between rebalancing and shutdown. I would agree that keeping consistency and behaving in the same way would be less error prone.

Reply:
Though is this going to be true in a multi processes environment? If I remember correctly you will have a separate process that collects the batches, writes to clickhouse and enqueues the offsets to commit. 
Do you already know how this is going to be synchronized with the calls to close() ? Are we just going to immediately drop the queue and ask the process to terminate and discard all the work that is in flight? 
I think it should be fine but we will have a higher risk of double writing to clickhouse. If we close everything between the write to clickhouse and the commit.